apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/FlushableConsumerBase.cs

```diff
@@ -20,87 +20,95 @@
-        protected readonly SemaphoreSlim Semaphore = new SemaphoreSlim(1);
+        private readonly SemaphoreSlim _semaphore = new SemaphoreSlim(1);
 
-        /// <summary>The consumer to flush to.</summary>
-        protected readonly IConsumer<T> Consumer;
-
-        /// <summary>The buffered items.</summary>
-        protected List<T> Items = new List<T>();
+        /// <summary>
+        /// Accepts an enumerable of items while the semaphore is held.
+        /// </summary>
+        /// <param name="items">The items to receive. Cannot be null.</param>
+        protected abstract void ReceiveWithSemaphoreHeld(IEnumerable<T> items);
 
-        public FlushableConsumerBase(IConsumer<T> consumer)
-        {
-            Consumer = GaxPreconditions.CheckNotNull(consumer, nameof(consumer));
-        }
+        /// <summary>
+        /// Accepts an enumerable of items asynchronously while the semaphore is held.
+        /// </summary>
+        /// <param name="items">The items to receive. Cannot be null.</param>
+        /// <param name="cancellationToken">The token to monitor for cancellation requests.</param>
+        /// <returns>A task representing the asynchronous operation.</returns>
+        protected abstract Task ReceiveAsyncWithSemaphoreHeld(
+            IEnumerable<T> items, CancellationToken cancellationToken);
 
-        /// <inheritdoc />
-        public abstract void Receive(IEnumerable<T> items);
+        /// <summary>
+        /// Flushes the consumer while the semaphore is held.
+        /// </summary>
+        protected abstract void FlushWithSemaphoreHeld();
 
-        /// <inheritdoc />
-        public abstract Task ReceiveAsync(IEnumerable<T> items, CancellationToken cancellationToken);
+        /// <summary>
+        /// Flushes the consumer asynchronously while the semaphore is held.
+        /// </summary>
+        /// <param name="cancellationToken">The token to monitor for cancellation requests.</param>
+        /// <returns>A task representing the asynchronous operation.</returns>
+        protected abstract Task FlushAsyncWithSemaphoreHeld(CancellationToken cancellationToken);
 
         /// <inheritdoc />
-        public void Flush()
+        public void Receive(IEnumerable<T> items)
         {
-            Semaphore.Wait();
+            _semaphore.Wait();
             try
             {
-                FlushWithSemaphoreHeld();
+                ReceiveWithSemaphoreHeld(items);
             }
             finally
             {
-                Semaphore.Release();
+                _semaphore.Release();
             }
         }
 
         /// <inheritdoc />
-        public async Task FlushAsync(CancellationToken cancellationToken = default(CancellationToken))
+        public Task ReceiveAsync(IEnumerable<T> items, CancellationToken cancellationToken)
         {
-            await Semaphore.WaitAsync(cancellationToken);
+            _semaphore.Wait();
```

Comment:
You should still use `WaitAsync` here, so I would add back the `async` and `await`s.